test/NVMLIOGroupTest.cpp

```diff
@@ -0,0 +1,330 @@
+
+    for (int accel_idx = 0; accel_idx < num_accelerator; ++accel_idx) {
+        EXPECT_NO_THROW(nvml_io.write_control("NVML::FREQUENCY", GEOPM_DOMAIN_BOARD_ACCELERATOR, accel_idx, 1530000000));
+        EXPECT_NO_THROW(nvml_io.write_control("FREQUENCY_ACCELERATOR_CONTROL", GEOPM_DOMAIN_BOARD_ACCELERATOR, accel_idx, 1530000000));
+        EXPECT_NO_THROW(nvml_io.write_control("NVML::FREQUENCY_RESET", GEOPM_DOMAIN_BOARD_ACCELERATOR, accel_idx, 12345));
+        EXPECT_NO_THROW(nvml_io.write_control("NVML::POWER_LIMIT", GEOPM_DOMAIN_BOARD_ACCELERATOR, accel_idx, 200));
+        EXPECT_NO_THROW(nvml_io.write_control("POWER_ACCELERATOR_LIMIT", GEOPM_DOMAIN_BOARD_ACCELERATOR, accel_idx, 200));
+    }
+}
+
+TEST_F(NVMLIOGroupTest, read_signal_and_batch)
+{
+    const int num_accelerator = m_platform_topo->num_domain(GEOPM_DOMAIN_BOARD_ACCELERATOR);
+
+    std::vector<double> mock_freq = {1530, 1320, 420, 135};
+    std::vector<int> batch_idx;
+
+    for (int accel_idx = 0; accel_idx < num_accelerator; ++accel_idx) {
+        EXPECT_CALL(*m_device_pool, frequency_status_sm(accel_idx)).WillRepeatedly(Return(mock_freq.at(accel_idx)));
+    }
+
+    NVMLIOGroup nvml_io(*m_platform_topo, *m_device_pool);
+
+    for (int accel_idx = 0; accel_idx < num_accelerator; ++accel_idx) {
+        batch_idx.push_back(nvml_io.push_signal("NVML::FREQUENCY", GEOPM_DOMAIN_BOARD_ACCELERATOR, accel_idx));
+    }
+    nvml_io.read_batch();
+
+    for (int accel_idx = 0; accel_idx < num_accelerator; ++accel_idx) {
+        double frequency = nvml_io.read_signal("NVML::FREQUENCY", GEOPM_DOMAIN_BOARD_ACCELERATOR, accel_idx);
+        double frequency_batch = nvml_io.sample(batch_idx.at(accel_idx));
+
+        EXPECT_DOUBLE_EQ(frequency, mock_freq.at(accel_idx)*1e6);
+        EXPECT_DOUBLE_EQ(frequency, frequency_batch);
+    }
+}
+
+TEST_F(NVMLIOGroupTest, read_signal)
+{
+    const int num_accelerator = m_platform_topo->num_domain(GEOPM_DOMAIN_BOARD_ACCELERATOR);
+    const int num_cpu = m_platform_topo->num_domain(GEOPM_DOMAIN_CPU);
+
+    std::vector<double> mock_freq = {1530, 1320, 420, 135};
+    std::vector<double> mock_utilization_accelerator = {100, 90, 50, 0};
+    std::vector<double> mock_power = {153600, 70000, 300000, 50000};
+    std::vector<double> mock_power_limit = {300000, 270000, 300000, 250000};
+    std::vector<double> mock_freq_mem = {877, 877, 877, 877};
+    std::vector<double> mock_throttle_reasons = {0, 1, 3, 128};
+    std::vector<double> mock_temperature = {45, 60, 68, 92};
+    std::vector<double> mock_energy = {630000, 280000, 470000, 950000};
+    std::vector<double> mock_performance_state = {0, 2, 3, 5};
+    std::vector<double> mock_pcie_rx_throughput = {4000, 3000, 2000, 0};
+    std::vector<double> mock_pcie_tx_throughput = {2000, 3000, 4000, 100};
+    std::vector<double> mock_utilization_mem = {25, 50, 100, 75};
+
+    /// @todo: Read /proc/sys/kernel/pid_max to guarantee that we never hit an actual PID in testing?
```

Comment:
In the initial test case I believe there would have been a failure since we wouldn't see the expected result when calling NVML::CPU_ACCELERATOR_ACTIVE_AFFINITIZATION, but that currently only EXPECTS_NO_THROW, so this is likely no longer necessary